pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+
+        /// A stable identifier for a collator.
+        type CollatorId: Member
+            + Parameter
+            + Ord
+            + MaybeSerializeDeserialize
+            + MaxEncodedLen
+            + TryFrom<Self::AccountId>;
+
+        /// The maximum number of sessions for which a collator can be inactive
+        /// before being moved to the offline queue
+        #[pallet::constant]
+        type MaxInactiveSessions: Get<u32>;
+
+        /// The maximum amount of collators that can stored for a session
+        #[pallet::constant]
+        type MaxCollatorsPerSession: Get<u32>;
+
+        /// Helper that returns the current session index.
+        type CurrentSessionIndex: GetSessionIndex<SessionIndex>;
+
+        /// Helper that returns the block author for the orchestrator chain (if it exists)
+        type GetSelfChainBlockAuthor: MaybeSelfChainBlockAuthor<Self::CollatorId>;
+
+        /// The weight information of this pallet.
+        type WeightInfo: weights::WeightInfo;
+    }
+
+    /// Switch to enable/disable activity tracking
+    #[pallet::storage]
+    pub type CurrentActivityTrackingStatus<T: Config> =
+        StorageValue<_, ActivityTrackingStatus, ValueQuery>;
+
+    /// A list of double map of active collators for a session
+    #[pallet::storage]
+    pub type ActiveCollators<T: Config> = StorageMap<
+        _,
+        Twox64Concat,
+        SessionIndex,
+        BoundedBTreeSet<T::CollatorId, T::MaxCollatorsPerSession>,
+        ValueQuery,
+    >;
+
+    /// A list of active collators for a session. Repopulated at the start of every session
+    #[pallet::storage]
+    pub type ActiveCollatorsForCurrentSession<T: Config> =
+        StorageValue<_, BoundedBTreeSet<T::CollatorId, T::MaxCollatorsPerSession>, ValueQuery>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Event emitted when the activity tracking status is updated
+        ActivityTrackingStatusSet { status: ActivityTrackingStatus },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The size of a collator set for a session has already reached MaxCollatorsPerSession value
+        MaxCollatorsPerSessionReached,
+        /// Error returned when the activity tracking status is attempted to be updated before the end session
+        ActivityStatusUpdateSuspended,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::set_inactivity_tracking_status())]
+        pub fn set_inactivity_tracking_status(
+            origin: OriginFor<T>,
+            is_enabled: bool,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            let current_status_end_session_index = match <CurrentActivityTrackingStatus<T>>::get() {
+                ActivityTrackingStatus::Enabled { start: _, end } => end,
+                ActivityTrackingStatus::Disabled { end } => end,
```

Comment:
I am guessing if this is disabled, there is no reason to re-disable it again no? so maybe we should early return in that case?

Reply:
Very good point! Also re-enabling after being enabled will stop recording activity for the current session. Added check that prevents that from happening + test in https://github.com/moondance-labs/tanssi/pull/922/commits/dfbcf46af9922f42291558f55de584f1419cc270